chore(broker): unify snapshot/DB services

MiguelPires opened the pull request:
closes #2353 
closes #1729

MiguelPires commented on logstreams/src/main/java/io/zeebe/logstreams/state/StateSnapshotController.java:
```diff
@@ -184,7 +184,7 @@ public long recover() throws Exception {
 
   @Override
-  public ZeebeDb openDb() {
+  public synchronized ZeebeDb openDb() {
```
If the exporter service and the partition both open the call openDb at the same time, they fail with a RocksDbException stating that no LOCK file is available

Zelldon replied:
Please just give the exporter only the Database. They don't need the snapshot controller anymore. Also remove the snapshoting related properties from the context.

MiguelPires commented on engine/src/main/java/io/zeebe/engine/processor/StreamProcessorController.java:
```diff
@@ -232,33 +203,6 @@ protected void onActorCloseRequested() {
-          () -> {
-            if (asyncSnapshotDirector != null) {
-              actor.runOnCompletionBlockingCurrentPhase(
-                  asyncSnapshotDirector.enforceSnapshotCreation(
```
Yes, in the `AsyncSnapshotDirector#close`

MiguelPires commented on engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotingDirectorService.java:
```diff
@@ -0,0 +1,77 @@
+import java.time.Duration;
+
+public class AsyncSnapshotingDirectorService implements Service<AsyncSnapshotingDirectorService> {
+  private final Injector<StreamProcessorService> streamProcessorServiceInjector = new Injector<>();
```
I tried it but I have to partially initialize the Director on creation and then set the stream processor controller things on service start. That makes it harder to test (we can make the stream process service and inject it manually but still it had some errors that didn't seem obvious) and it wasn't a very elegant solution to have the initialization split because of the dependencies. But I agree that in services without injectors (like the LogBlockIndexService) it would be a nice refactor

MiguelPires commented on broker-core/src/test/java/io/zeebe/broker/exporter/stream/ExporterDirectorTest.java:
```diff
@@ -327,7 +336,7 @@ public void shouldUpdateLastExportedPositionOnClose() {
 
   @Test
-  public void shouldRemoveExporterFromState() {
+  public void shouldRemoveExporterFromState() throws Exception {
```
I added a `zeebeDb.close` to `closeExporterDirector()` because of a bug where the lock file wasn't released which would cause an exception when restarting the exporter director.

Zelldon replied:
Ah ok thanks

MiguelPires commented on broker-core/src/test/java/io/zeebe/broker/exporter/ExporterManagerTest.java:
```diff
@@ -82,33 +84,47 @@ public void shouldRestoreExporterFromState() {
   public void shouldRemoveExporterFromState() {
-
     // given
+    waitForStreamProcessor();
```
It's not, I removed it in the latest commit

Zelldon replied:
And the other wait statements? Maybe we can adjust the hack in the broker which already waits on the partition?

MiguelPires commented:
Thanks for the feedback, I made some new changes with your suggestions

MiguelPires commented on broker-core/src/test/java/io/zeebe/broker/exporter/ExporterManagerTest.java:
```diff
@@ -82,33 +84,47 @@ public void shouldRestoreExporterFromState() {
+    // wait for service to be installed
+    waitForStreamProcessor();
+
+    // TODO: remove workaround to force new snapshot by publishing new record
```
I agree, It's an extreme edge case =P

Zelldon commented on engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotDirector.java:
```diff
@@ -234,14 +229,17 @@ private void onCommitCheck() {
                           if (ex2 == null) {
                             enforceSnapshotCreation(writtenPosition, processedPosition);
+                            metrics.close();
                             actor.close();
```
We probably want to close the actor and metrics anyway, even if the snapshot was not successful ? What do you think

MiguelPires commented:
Hey @Zelldon, I applied your hints. The last commit might seem a little strange but in the case where the StreamProcessorController actor is still on the `onActorStarted` method and hasn't initialized the ProcessingStateMachine when the services starts closing, a NPE would occur because the snapshot forcing uses the state machine to get the processed and written positions. This way those calls are scheduled after the initialization.

Zelldon commented on engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotingDirectorService.java:
```diff
@@ -73,7 +73,7 @@ public void start(final ServiceStartContext startContext) {
   public void stop(final ServiceStopContext stopContext) {
     if (asyncSnapshotDirector != null) {
-      stopContext.async(asyncSnapshotDirector.close());
+      stopContext.async(asyncSnapshotDirector.asyncClose());
```
Normally we use the term `closeAsync`

Zelldon commented on engine/src/main/java/io/zeebe/engine/processor/StreamProcessorController.java:
```diff
@@ -250,11 +250,17 @@ private void resume() {
-    return processingStateMachine.getLastWrittenPositionAsync();
+    final CompletableActorFuture<Long> future = new CompletableActorFuture();
+    actor.call(
+        () -> future.complete(processingStateMachine.getLastSuccessfulProcessedEventPosition()));
```
Should be last written right?

Zelldon replied:
Maybe we need a test for this :laughing: :see_no_evil:

Zelldon commented on engine/src/main/java/io/zeebe/engine/processor/StreamProcessorController.java:
```diff
@@ -250,11 +250,17 @@ private void resume() {
   public ActorFuture<Long> getLastWrittenPositionAsync() {
-    return processingStateMachine.getLastWrittenPositionAsync();
+    final CompletableActorFuture<Long> future = new CompletableActorFuture();
+    actor.call(
```
Actor.call returns a future with the callable return value, so you can use this instead of a separate future

MiguelPires commented:
I merged this branch with the other refactoring. I had added a test for the last processed and last written but it didn't translate well into the new tests (also it used the suspend feature) so I removed it, I hope that's ok

Zelldon commented:
bors r+

Zelldon commented:
bors r+

Zelldon commented:
bors r+

Zelldon commented:
I hate this
bors r+  :construction_worker_man:

menski commented:
one more time
bors r+

MiguelPires commented:
bors r+

Zelldon commented:
bors r+
